diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -2747,20 +2747,9 @@ void maybe_intro_message(void)
 void intro_message(int colon)
 {
   static char *(lines[]) = {
-    N_(NVIM_VERSION_LONG),
+    "I use Neovim (BTW)",
     "",
-    N_("Nvim is open source and freely distributable"),
-    "https://neovim.io/#chat",
-    "",
-    N_("type  :help nvim<Enter>       if you are new! "),
-    N_("type  :checkhealth<Enter>     to optimize Nvim"),
-    N_("type  :q<Enter>               to exit         "),
-    N_("type  :help<Enter>            for help        "),
-    "",
-    N_("type  :help news<Enter> to see changes in v%s.%s"),
-    "",
-    N_("Help poor children in Uganda!"),
-    N_("type  :help iccf<Enter>       for information "),
+    N_(NVIM_VERSION_MEDIUM),
   };
 
   // blanklines = screen height - # message lines

