--- a/lua/null-ls/helpers/cache.lua
+++ b/lua/null-ls/helpers/cache.lua
@@ -43,20 +43,23 @@
 M.by_bufnr_async = function(cb)
     -- assign next available key, since we just want to avoid collisions
     local key = next_key
-    M.cache[key] = {}
     next_key = next_key + 1

     return function(params, done)
+        -- RPC termination clears M.cache, but registered resolvers are reused.
+        local values = M.cache[key] or {}
+        M.cache[key] = values
+
         local bufnr = params.bufnr
         -- if we haven't cached a value yet, get it from cb
-        if M.cache[key][bufnr] == nil then
+        if values[bufnr] == nil then
             -- make sure we always store a value so we know we've already called cb
             cb(params, function(result)
-                M.cache[key][bufnr] = result or false
-                done(vim.deepcopy(M.cache[key][bufnr]))
+                values[bufnr] = result or false
+                done(vim.deepcopy(values[bufnr]))
             end)
         else
-            done(vim.deepcopy(M.cache[key][bufnr]))
+            done(vim.deepcopy(values[bufnr]))
         end
     end
 end
@@ -67,21 +70,24 @@
 M.by_bufroot_async = function(cb)
     -- assign next available key, since we just want to avoid collisions
     local key = next_key
-    M.cache[key] = {}
     next_key = next_key + 1

     return function(params, done)
+        -- RPC termination clears M.cache, but registered resolvers are reused.
+        local values = M.cache[key] or {}
+        M.cache[key] = values
+
         local root = params.root
         assert(root, "root should not be empty when caching by bufroot")
         -- if we haven't cached a value yet, get it from cb
-        if M.cache[key][root] == nil then
+        if values[root] == nil then
             -- make sure we always store a value so we know we've already called cb
             cb(params, function(result)
-                M.cache[key][root] = result or false
-                done(vim.deepcopy(M.cache[key][root]))
+                values[root] = result or false
+                done(vim.deepcopy(values[root]))
             end)
         else
-            done(vim.deepcopy(M.cache[key][root]))
+            done(vim.deepcopy(values[root]))
         end
     end
 end
